### ISSUE TITLE
[FR] Enhance text effect

### ISSUE THREAD

**Issue by oc1024** [olive-editor/olive#523]
__Olive version:__  3d0a65c

Currently the Render > Text tool is very limited in Olive.

It can:
* Add text, set its color and size.
* Add outline, set its color and size.
* Add shadow, set its color and soften it.

However:
* Can't set shadow angle.
* Can't activate bold, italic, underline, strike-through. With some effort it's possible to recreate bold (spacing suffers), underline and strike-through, however, italic is simply impossible to do.

Personally, this is the extent of my complaints.

Some things to think of in the long run:
* Full Rich Editing support (multiple fonts, multiple settings on the same effect)
* Advanced strike-through/underline (set thickness and position)

**Comment by Symbian9** [comment]
> Can't set shadow angle.

This already requested in #460

**Comment by oc1024** [comment]
> This already requested in #460

You mean #490?

**Comment by Symbian9** [comment]
> You mean 490

Yeah, just fixed my comment

**Comment by NathanSalapat** [comment]
Adding text is also bound to the project 'canvas' size, so currently scrolling text is impossible. I ended up just creating an image with the text on a transparent background and animated it moving, but it would be nice if we could use the internal text tool to do credit rolls.

**Comment by Symbian9** [comment]
> so currentl scrolling text is impossible.

TL;DR: Scrolling text possible using "Transform" and separating text in few slides stacked in row.

P.S.: I will add example today.

**Comment by mdmayfield** [comment]
@Symbian9 I can imagine exactly what you have in mind. That's a good idea for a temporary workaround and definitely makes scrolling *possible*, though not easy or convenient.

Similarly, to include text of different fonts, colors, sizes, etc., you can "stack" text effects. It's *possible*; the functionality is already there, in a sense.

But I believe that, in the long term, expecting this sort of approach on the part of the user would reduce Olive's chances of being seen as a truly professional tool.

Often, my colleagues who also work with digital media have expressed disdain for open source media tools. I believe many open source applications have a bad reputation not because they lack *functionality*, but because they are markedly less *convenient* to use than proprietary ones.

That is, for many workflows, such as this scrolling technique, they put the burden of implementing useful functionality on the *user* rather than on the *software*. Ergonomics and convenience can go a very long way in increasing the usefulness and popularity of an application.

**Comment by itsmattkc** [comment]
Check out the rich text effect in the latest commit for most of the issues in this thread addressed (except strikethrough which QTextEdit's HTML transfer doesn't seem to support).

**Comment by Symbian9** [comment]
> Check out the rich text effect in the latest commit

FTR, latest AppImage builds not available

**Comment by mdmayfield** [comment]
This is fantastic! @itsmattkc Thank you for implementing this.

I'm seeing a segfault when right-clicking and choosing "Edit Text" in a regular non-rich text field - will file a separate issue for that.

Eventually, for a total pro-level workflow, I'd suggest a long-term goal to have the text editing controls inline with the rest of the effect rather than in a separate dialog, and to have the rich text live-update in the Sequence Viewer (the way the non-rich text editing works now).

Great to see this progressing! This is already a huge improvement and is making it much easier to create titles with highlighted words, or references with italics, etc.

**Comment by Symbian9** [comment]
> Eventually, for a total pro-level workflow, I'd suggest a long-term goal to have the text editing controls inline with the rest of the effect rather than in a separate dialog, and to have the rich text live-update in the Sequence Viewer (the way the non-rich text editing works now).

For pro-level workflow need redesign text effect and **add new clip type: "Text Clip"**

**Comment by Symbian9** [comment]
> Check out the rich text effect  ... QTextEdit's HTML

Could you give us text with markup example of such effect?

**Comment by unfa** [comment]
I'd also want to add:
Character, Word and Line separation settings would be great to have :)

**Comment by Symbian9** [comment]
Well, I finally found ***"Rich Text" effect*** ;-)

@itsmattkc But there are few small notes/issues

- "Vertical Align" should be active only if "Auto-Scroll" set to "Off";
- all shadow option not connected to "Shadow" state;
- iff possible, add button `¶` for "Toggle Formating Marks" into text editor toolbar;
- add "Apply" button in "Edit Text" window (for change text without closing "Edit Text" widow)
- show button "Edit Text" if user set "Effect Textbox Lines" to `0` in "Preferences" (for "Text" and "Rich Text")

![pic.1](https://i.imgur.com/9kgA1XZ.png)

**Comment by Symbian9** [comment]
> Character, Word and Line separation settings would be great to have :)

Yeah. "Line Wraping" option for "Rich Text" really needed option

**Comment by Symbian9** [comment]
> - show button "Edit Text" if user set "Effect Textbox Lines" to 0 in "Preferences" (for "Text" and "Rich Text")

Now its partially solved, but add option for hide input field if set "Effect Textbox Lines" to 0

> - all shadow option not connected to "Shadow" state;

Still not fixed

![pic.1](https://i.imgur.com/ZgSS4P0.png)

Additionally there some extra indents around input field

![pic.2](https://i.imgur.com/KqKGi8f.png)

@itsmattkc also take a look on this issues

> - "Vertical Align" should be active only if "Auto-Scroll" set to "Off";
> - iff possible, add button ¶ for "Toggle Formating Marks" into text editor toolbar;
add "Apply" button in "Edit Text" window (for change text without closing "Edit Text" widow)

**Comment by oliwerejder** [comment]
How do i add outline to text and where do i ask support for olive if there is any?

**Comment by benrob0329** [comment]
Adding outline support would be nice.

**Comment by Symbian9** [comment]
FTR, @oliwerejder @benrob0329 , "Text" effect (not "Rich Text") already has text outline support.

**Comment by benrob0329** [comment]
I suppose I don't understand a couple things.

1. Why we have two text effects
2. Why "Rich Text" lacks a style option that non-rich text has.

**Comment by itsmattkc** [comment]
@benrob0329 Ultimately I agree, the text effects are fragmented and should be merged into one. The reason it's currently like this is because the standard "text" effect is achieved by adding text to a QPainterPath which can easily be drawn with a fill and a stroke, whereas the "rich text" is achieved with a QTextDocument which cannot be converted to a path in the same way, making a stroke much more difficult. There are also other limitations with the "rich text" approach, such as no ability to keyframe font size, color, etc.

The solution I think will come with the upcoming node rewrite and breaking both effects down into base components. A simple text generator where the font can be keyframed, a rich text generator that allows multiple formatting, and then connecting those to separate outline, shadow, etc. nodes if desired, rather than having two effects that duplicate a lot of each others' functionality. These outline/shadow nodes could also be used for any non-text clip as well.

**Comment by N6REJ** [comment]
How in the world do you even add text?  I don't see it anywhere.

**Comment by Masken8** [comment]
Click the *plus icon* to the *left* of the *timeline *and then* text*

fre 19 apr. 2019 kl. 04:08 skrev Bear <notifications@github.com>:

**Comment by itsmattkc** [comment]
This will be addressed in the node rewrite.
